src/transform/src/analysis/equivalences.rs

```diff
@@ -477,12 +477,148 @@ impl EquivalenceClasses {
+            }
+
+            // Termination detection.
+            if prev != self.remap {
+                previous = Some(self.remap.clone());
+            } else {
+                previous = None;
+            }
+        }
+    }
+
+    /// Invoke `self.implications()` to propose equivalences, and then judge them to be novel or not.
+    ///
+    /// The method may add new classes to `self.classes`, and will return true exactly in that case.
+    /// This should eventually return `false` once `self.implications()` returns no *novel* classes,
+    /// defined as classes that when reduced (under `reduce_child`) result in at least two distinct
+    /// expressions, that do not all already have the same representative in `self.remap`.
+    ///
+    /// This method may produce non-novel equivalences, due to its inability to perform `MSE::reduce`.
+    /// We can end up with e.g. constant expressions that cannot be found until they are so reduced.
+    fn expand(&self) -> Vec<Vec<MirScalarExpr>> {
+        // Consider expanding `self.classes` with novel equivalences.
+        let mut novel = self.implications();
+        for class in novel.iter_mut() {
+            // reduce each expression to its canonical form.
+            for expr in class.iter_mut() {
+                self.remap.reduce_expr(expr);
+            }
+            class.sort();
+            class.dedup();
+            // for a class to be interesting we require at least two elements that do not reference the same root.
+            let common_class = class
+                .iter()
+                .map(|x| self.remap.get(x))
+                .reduce(|prev, this| if prev == this { prev } else { None });
+            if class.len() == 1 || common_class != Some(None) {
+                class.clear();
+            }
+        }
+        novel.retain(|c| !c.is_empty());
+        novel
+    }
+
+    /// Derives potentially novel equivalences without regard for minimization.
+    ///
+    /// This is an opportunity to explore equivalences that do not correspond to expression minimization,
+    /// and therefore should not be used in `minimize_once`. They are still potentially important, but
+    /// required additional guardrails to ensure we reach a fixed point.
+    ///
+    /// The implications will be introduced into `self.classes` and will prompt a round of minimization,
+    /// making it somewhat polite to avoid producing outputs that cannot result in novel equivalences.
+    /// For example, before producing a new equivalence, one could check that the involved terms are not
+    /// already present in the same class.
+    fn implications(&self) -> Vec<Vec<MirScalarExpr>> {
+        let mut new_equivalences = Vec::new();
+
+        // If we see `false == IsNull(foo)` we can add the non-null implications of `foo`.
+        let mut non_null = std::collections::BTreeSet::default();
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_false()) {
```

Comment:
We're searching every class to find the (at most) one class that has the literal false in it. We could instead force the invariant that there is always such a class and look it up directly. (Something egraphs do well is let you ask, "what is the class for term `foo`?".)

Reply:
I agree, and really wish we had e-graphs! :D

Baking a class for `false` could work, but you end up with some grot because .. well you probably want one for `true` also, and maybe they are the same (what?!?), and you have to remember to always look for it, and migrate everything there if anything reduces to false. Idk, seems reasonable, but a bunch of rewriting. Mostly tempted to reconsider that around the same time as thinking about e-graphs proper. Unless there's a low friction way to slot it in.

One thing that could make this lightly easier is only looking at the prefix of the class that is literals, and bailing as soon as that range ends.

Reply:
In person discussion: this moment, unlike several others, feels especially bad because we aren't otherwise going to do a linear pass over all expressions, and so searching everything is potentially a dramatic increase in cost. We discussed a few light alternatives, including e.g.
1. only scan through the literal prefix of each class (early exit on long classes)
2. sort the classes by complexity, so we are likely to see literals early and can bail out of class inspection (early exit on many classes).
3. maintain notes about where to find various literals (frank dislikes for now, on account of ability to get things wrong).

Reply:
The first option has been pushed, and while I like the second option it's awkward at the moment because while the classes are sorted within by complexity, they are sorted amongst each other by `Ord::le`, which prefers columns to literals. T.T  Fixable, but .. I'm more inclined to fix `Ord::le` to be what we want and smoke out errors there.